napalm/junos/junos.py

```diff
@@ -235,14 +237,75 @@ def compare_config(self):
             return diff.strip()
 
     def commit_config(self, confirmed=None, message=None):
-        """Commit configuration."""
-        if confirmed is not None:
-            raise NotImplementedError
+        """
+        Commit configuration.
+
+        if confirmed is passed, random string is generated for the commit log
+        """
+        pending_commit_string = None
+        if confirmed:
+            pending_commit_string = str(uuid.uuid4()).split('-')[0]
+            if message:
+                message = '{}_napalm_confirm_{}'.format(message, pending_commit_string)
+            else:
+                message = 'napalm_confirm_{}'.format(pending_commit_string)
 
-        self.device.cu.commit(ignore_warning=self.ignore_warning)
+        self.device.cu.commit(
+            ignore_warning=self.ignore_warning,
+            confirm=confirmed,
+            comment=message
+        )
+        self._pending_commit_string = pending_commit_string
         if not self.config_lock:
             self._unlock()
 
+    def commit_confirm(self):
+        """
+        Confirms a commit
+
+        Simply sends a commit to confirm a commit.
+
+        Will send the commit even without a pending
+        """
+        if self.has_pending_commit_confirm:
+            self.commit_config()
+            return
+        log.warning('no pending commit to confirm on %s', self.hostname)
+
+    @property
+    def has_pending_commit_confirm(self):
```

Comment:
Relying on an object attribute for a non locked, non atomic transaction seems brittle to me.

Reply:
I guess it depends if we want to return if there is a confirm pending in the system or by napalm. So I guess that's something we should agree on first and add it to documentation as a note.

Reply:
I think we cannot realistically aim to return if there's a confirm pending in the system, but limit to napalm only: it is doable for Junos, but for the other platforms (particularly IOS and NXOS) it might be impossible and we'd aim way too far. I would put the following example

Junos
------

1. Commit confirmed some changes
2. Another user from the CLI commits some changes (bad!!!)
3. You commit is not confirmed and it will revert. With your current implementation it would revert *only* your changes. Which I agree is the right thing to do.

IOS/Other stupid platform
--------------------------

1. Commit confirmed some changes
2. Another user from the CLI commits some changes (bad!!!)
3. I don't think it's possible to revert *only* your changes, but instead we might revert the changes made from a different process / session as well. If that's the case (@ktbyers to confirm here please), in order to keep the drivers in-line, we might need to make a compromise.

Reply:
Not sure the problem you are describing is really relevant or not. I understand it's a shitty problem for IOS but the question is if `commit_confirm`, `commit_confirm_revert` and `has_pending_commit` should work at the system level or not:

1. If it works at the system level it is the user responsibility to make sure it's done under controlled circumstances because in that case those methods are just going to blindly run the commands for those operations in the system.
2. If it works under napalm only same as before but we do a check like `self.pending_commit`.

The problem of option 2 is that this is not going to work with ansible or if your connection drops as you'd lose that variable.

Reply:
Are we actually able to implement (1) properly on IOS/NXOS?

Reply:
That I don't know and I will defer to our resident expert @ktbyers but if we can't I think this feature is going to be useless for most of our users and dangerous for other. I'd certainly don't use it if I can't recover from a connection drop and can confirm it.

Reply:
I will try to review this thread sometime today. I believe I had a reasonable solution for IOS, NX-OS, and EOS (with EOS I think being the most problematic IIRC). 

I need to read your comments though to see how the solutions addresses the points brought up.

Reply:
EOS should have native support for the commit confirm operations I think.

Reply:
Just adding comments as I read through this thread:
1. IMO it should be system commit-confirm not a NAPALM specific commit-confirm
2. It is already done on IOS (i.e. this branch has a solution implemented in IOS that does a system commit-confirm).
3. I will have to check the solution I had in mind for NX-OS, but I am pretty sure I came to the conclusion a reasonable solution could be implemented on NX-OS.
4. IIRC the EOS solution requires a certain version of EOS which I have noted somewhere (or can figure it out again).
5. I agree with @dbarrosop point here that we should require this (I think this is consistent with what is realistically already required).

```
If it works at the system level it is the user responsibility to make sure 
it's done under controlled circumstances because in that case those 
methods are just going to blindly run the commands for those operations
 in the system.
```

6. The solution should survive a NAPALM connection drop / re-establishment.